Data race in ParallelBlobDownloader

Go test failes on kv branch (3185b62) and I got this error:

```
WARNING: DATA RACE
Write at 0x00c00038a2e8 by goroutine 126:
  lukechampine.com/us/renterhost.(*Session).setErr()
      /Users/junpei/src/github.com/lukechampine/us/renterhost/session.go:77 +0xbd
  lukechampine.com/us/renterhost.(*Session).readMessage()
      /Users/junpei/src/github.com/lukechampine/us/renterhost/session.go:153 +0xb6b
  lukechampine.com/us/renterhost.(*Session).ReadResponse()
      /Users/junpei/src/github.com/lukechampine/us/renterhost/session.go:224 +0x139
  lukechampine.com/us/host.(*session).readResponse()
      /Users/junpei/src/github.com/lukechampine/us/host/session.go:84 +0x13c
  lukechampine.com/us/host.(*SessionHandler).rpcRead.func1()
      /Users/junpei/src/github.com/lukechampine/us/host/session.go:518 +0xa4

Previous read at 0x00c00038a2e8 by goroutine 118:
  lukechampine.com/us/renterhost.(*Session).writeMessage()
      /Users/junpei/src/github.com/lukechampine/us/renterhost/session.go:113 +0x66
  lukechampine.com/us/renterhost.(*Session).WriteResponse()
      /Users/junpei/src/github.com/lukechampine/us/renterhost/session.go:216 +0x184
  lukechampine.com/us/host.(*session).writeResponse()
      /Users/junpei/src/github.com/lukechampine/us/host/session.go:104 +0x150
  lukechampine.com/us/host.(*SessionHandler).rpcRead()
      /Users/junpei/src/github.com/lukechampine/us/host/session.go:594 +0xe64
  lukechampine.com/us/host.(*SessionHandler).rpcRead-fm()
      /Users/junpei/src/github.com/lukechampine/us/host/session.go:506 +0x4b
  lukechampine.com/us/host.(*SessionHandler).Serve()
      /Users/junpei/src/github.com/lukechampine/us/host/session.go:233 +0x7b4
  lukechampine.com/us/internal/ghost.listen.func1()
      /Users/junpei/src/github.com/lukechampine/us/internal/ghost/ghost.go:126 +0xa9

Goroutine 126 (running) created at:
  lukechampine.com/us/host.(*SessionHandler).rpcRead()
      /Users/junpei/src/github.com/lukechampine/us/host/session.go:516 +0x15b
  lukechampine.com/us/host.(*SessionHandler).rpcRead-fm()
      /Users/junpei/src/github.com/lukechampine/us/host/session.go:506 +0x4b
  lukechampine.com/us/host.(*SessionHandler).Serve()
      /Users/junpei/src/github.com/lukechampine/us/host/session.go:233 +0x7b4
  lukechampine.com/us/internal/ghost.listen.func1()
      /Users/junpei/src/github.com/lukechampine/us/internal/ghost/ghost.go:126 +0xa9

Goroutine 118 (running) created at:
  lukechampine.com/us/internal/ghost.listen()
      /Users/junpei/src/github.com/lukechampine/us/internal/ghost/ghost.go:124 +0x5c
==================
--- FAIL: TestKVCancel (3.80s)
    testing.go:1038: race detected during execution of test
```

--- a/renterhost/session.go
+++ b/renterhost/session.go
@@ -12,6 +12,7 @@ import (
 	"io/ioutil"
 	"net"
 	"strings"
+	"sync"
 
 	"github.com/aead/chacha20/chacha"
 	"github.com/pkg/errors"
@@ -65,12 +66,16 @@ type Session struct {
 	inbuf     objBuffer
 	outbuf    objBuffer
 	challenge [16]byte
-	err       error // set when Session is prematurely closed
-	closed    bool
 	isRenter  bool
+
+	mu     sync.Mutex
+	err    error // set when Session is prematurely closed
+	closed bool
 }
 
 func (s *Session) setErr(err error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
 	if err != nil && s.err == nil {
 		if ne, ok := err.(net.Error); !ok || !ne.Temporary() {
 			s.conn.Close()
@@ -81,11 +86,19 @@ func (s *Session) setErr(err error) {
 
 // PrematureCloseErr returns the error that resulted in the Session being closed
 // prematurely.
-func (s *Session) PrematureCloseErr() error { return s.err }
+func (s *Session) PrematureCloseErr() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.err
+}
 
 // IsClosed returns whether the Session is closed. Check PrematureCloseErr to
 // determine whether the Session was closed gracefully.
-func (s *Session) IsClosed() bool { return s.closed || s.err != nil }
+func (s *Session) IsClosed() bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.closed || s.err != nil
+}
 
 // SetChallenge sets the current session challenge.
 func (s *Session) SetChallenge(challenge [16]byte) {
@@ -110,8 +123,8 @@ func (s *Session) VerifyChallenge(sig []byte, pub ed25519.PublicKey) bool {
 }
 
 func (s *Session) writeMessage(obj ProtocolObject) error {
-	if s.err != nil {
-		return s.err
+	if err := s.PrematureCloseErr(); err != nil {
+		return err
 	}
 	// generate random nonce
 	nonce := make([]byte, 256)[:s.aead.NonceSize()] // avoid heap alloc
@@ -142,8 +155,8 @@ func (s *Session) writeMessage(obj ProtocolObject) error {
 }
 
 func (s *Session) readMessage(obj ProtocolObject, maxLen uint64) error {
-	if s.err != nil {
-		return s.err
+	if err := s.PrematureCloseErr(); err != nil {
+		return err
 	}
 	if maxLen < MinMessageSize {
 		maxLen = MinMessageSize
@@ -347,12 +360,14 @@ func (s *Session) RawResponse(maxLen uint64) (*ResponseReader, error) {
 // Close gracefully terminates the RPC loop and closes the connection.
 func (s *Session) Close() (err error) {
 	defer wrapErr(&err, "Close")
-	if s.closed || s.err != nil {
+	if s.IsClosed() {
 		return nil
 	}
+	s.mu.Lock()
 	s.closed = true
+	s.mu.Unlock()
 	if s.isRenter {
-		s.WriteRequest(loopExit, nil)
+		s.writeMessage(&loopExit)
 	}
 	return s.conn.Close()
 }

